RecoDataProducts/inc/STMWaveformCollection.hh

```diff
@@ -0,0 +1,28 @@
+#ifndef RecoDataProducts_STMWaveformCollection_hh
+#define RecoDataProducts_STMWaveformCollection_hh
+//
+// The STMWaveformCollection isn't just a vector of STMWaveforms.
+// It contains a channel number since we plan to keep the waveforms from each detector separate
+//
+
+#include "Offline/RecoDataProducts/inc/STMWaveform.hh"
+
+namespace mu2e {
+
+  class STMWaveformCollection {
+  public:
+    STMWaveformCollection() : _channel(STMChannel::unknown) {}
+    STMWaveformCollection(STMChannel ch) : _channel(ch) {}
+
+    STMChannel channel() const { return _channel; }
+
+    const size_t size() const { return _waveforms.size(); }
+    void push_back(const mu2e::STMWaveform& waveform) { _waveforms.push_back(waveform); }
+    const std::vector<mu2e::STMWaveform>& waveforms() const { return _waveforms; }
+
+  private:
+    mu2e::STMChannel _channel;
```

Comment:
Is it possible to remove this channel member and operate by identifying the channel using the "instance" field of the product name?

Reply:
It's definitely possible, and I'm open to changing it. My worry with identifying the channel with the instance name is that a user could misconfigure later modules.

Just to check I understand your proposal with an example. At the moment, in a module that would calibrate the energies we would have something like this:
```
void MakeSTMHits::produce(art::Event& event) {
    auto waveformsHandle = event.getValidHandle<STMWaveformCollection>(_stmWaveformsTag);
    STMEnergyCalib const& stmEnergyCalib = _stmEnergyCalib_h.get(event.id()); // get calibration
    STMChannel ch = waveformsHandle ->channel();
    const auto& pars = stmEnergyCalib.calib(ch);
    for (const auto & waveform : waveformsHandle->waveforms()) { // I don't like the way this line look
        // do calibration
    }
}
```
(Admittedly we won't be calibrating the waveforms themselves but the object that will represent the pulse area).
And then your proposal would be something like this (changed lines highlighted with ```**```):
```
void MakeSTMHits::produce(art::Event& event) {
    auto waveformsHandle = event.getValidHandle<STMWaveformCollection>(_stmWaveformsTag);
    STMEnergyCalib const& stmEnergyCalib = _stmEnergyCalib_h.get(event.id()); // get calibration
**    STMChannel ch = STMChannel(_stmWaveformsTag.instance())
    const auto& pars = stmEnergyCalib.calib(ch);
**    for (const auto & waveform : *waveformsHandle) {
        // do calibration
    }
}
```
Is that right?

I was worried about misconfiguring modules when I thought we would need to pass in an extra fcl parameter and so a user could get their wires crossed. But I think this gets around that, right?

Reply:
Yes, you are demonstrating the idea I suggested.  Please note that I am not really promoting the suggestion, more just asking if it makes sense.  I'd have to understand your larger plans to see if I might develop a stronger opinion.

Reply:
Thanks, Ray. I like the suggestion. One thing I realized that might get a little wordy is in the collection names later on. For example, we will have a module like ```STMZeroSuppression``` and we will have two instances of it, one for each channel. So our fcl would look something like this:

```
zeroSuppressHPGe : { module_type: STMZeroSuppression
 # HPGe specific parameters
}
zeroSuppressLaBr : { module_type: STMZeroSuppression
 # LaBr specific parameters
}
```
If we put the channel name in the instance name, then we will have output collections ```zeroSuppressHPGe:HPGe``` and ```zeroSuppressLaBr:LaBr```, which is a little unsightly but not a big deal.

Let me know if this gives you a stronger opinion, otherwise I will implement your idea (I like that we can just have a vector of STMWaveforms) and we can revisit this in future PRs when there is more of the bigger picture in place.

Reply:
>If we put the channel name in the instance name, then we will have output collections zeroSuppressHPGe:HPGe and zeroSuppressLaBr:LaBr, which is a little unsightly but not a big deal.

this is even better - the collections are identified by the module name field of the product name, so as your example here indicated, adding an instance name is redundant and can be skipped.  Then this also satisfies Andrei's rule.   And the comment that the internal class field is not needed continues to be relevant.

Reply:
There are two challenges with idenitfying the channel from the module label and taking it out of the internal class field.

The first is that, to get the channel for e.g. the calibration constants, we would need to parse the module label. This isn't a big challenge (since we can write a utility that just looks at the last 4 characters) but I don't think it's ideal.

The second challenge is that, for the test beam data, I have written the source module to create collections for both detectors. I could rewrite it to only create one collection and have two fcl files (one for each detector). Then we would run over the raw HPGe or the raw LaBr data with the appropriate fcl file. This is fine for the test beam data since we only had one detector connected at a time so don't need to worry about storing data for both.  

Having said that, I don't know what the situation will be when taking data. In the online/otsdaq/artdaq world, will data products be written with different module labels? I'll have to find out...

Anyway, a potential compromise for the time being would be to identify the channel by the module label (most cases) *or* the instance name (if one exists like for the source module).

Reply:
I didn't mean to drag out this comment so much.  I don't really have an opinion, it was just a suggestion/observation.  Any time you are sure you are on the right track, you should just go with it.  I think for me to comment on your latest comments, I'd need to understand the whole processing plan.  If you would like me to do that, let me know.  I don't think I would have any unique insight, just another pair of eyes.  I'm happy to do that if it would help.

As far as I know, the trigger process will act like a normal art job and your products will belong to your module which reads fragments.  I think the datalogger can alter events, but I don't see why it would in this case. But you should  probably confirm.